modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ReferenceNode.java

```diff
@@ -17,26 +17,25 @@
  * under the License.
  */
 
-package org.elasticsearch.painless.node;
+package org.elasticsearch.painless.ir;
 
+import java.util.ArrayList;
 import java.util.List;
 
-/**
- * Interface for lambda/method reference nodes. They need special handling by LDefCall.
- * <p>
- * This is because they know nothing about the target interface, and can only push
- * all their captures onto the stack and defer everything until link-time.
- */
-interface ILambda {
+public abstract class ReferenceNode extends ExpressionNode {
+
+    /* ---- begin node data ---- */
 
-    /** Returns reference to resolve at link-time */
-    String getPointer();
+    private final List<String> captures = new ArrayList<>();
 
-    /** Returns the types of captured parameters. Can be empty */
-    List<Class<?>> getCaptures();
+    public void addCapture(String capture) {
+        captures.add(capture);
+    }
 
-    /** Returns the number of captured parameters */
-    default int getCaptureCount() {
-        return getCaptures().size();
+    public List<String> getCaptures() {
+        return captures;
```

Comment:
is it ok to leak a mutable data structure here?

Reply:
all ir nodes are designed to be completely mutable as we hope to modify the ir tree structure at some point to add optimizations and possibly other features; we keep the user tree as immutable and hide the ir tree behind it